### DISCUSSION TITLE
20230704 회의

### DISCUSSION THREAD

**Discussion by KKangTaee** [KKangTaee/HifeSurvival#24] [General]
## 확인 및 수정할 내용
* IncreaseStat 호출 시 올바르게 값을 전달받고 내려주는지 체크 (서버)
* #28
* 특정 몬스터 사망시 사라지는 연출 안나오는 이슈 (클라)
* #27 
* 상대방 캐릭터가 정지 했음에도 Idle모션으로 안 돌아오는 이슈 (클라)
* body range 충돌 작업 필요 (공동)
* 선택 씬에서 아이콘들 안 나옴 (클라)

## 해야할 일
1. #29
2. #30
3. #31 
4. #32 
5. 몬스터 공격 연출(클라)
   * 공격 시 파티클 처리 
7. 몬스터 이동 추가 폴리싱 (서버)
   * 몬스터 AStar 적용
8. 기타
   * 어드레서블 매니저 구현(추가다운로드) (클라)

### 아이템
* 플레이어가 아이템 풀장착인 상황
	* 추가 획득을 못하게 할 것인가? 골드로 변경해서 지급할 것인가? 아니면 아이템 경험치로 사용?
* 아이템 레벨업 https://github.com/KKangTaee/HifeSurvival/discussions/24#discussioncomment-6348992
	* 아이템에 경험치를 두어 레벨업 시스템으로? 아니면 그냥 스텟 증가처럼? (item_upgrade 게임데이터)
	* 아이템 정보 UI 를 만들어, 여기서 아이템에 대한 정보를 노출시키고 업그래이드 할것인지 팔것인지 처리를 한다.
* 아이템 사용
	* 어떤식으로 패킷을 주고 받을까?
	* 공격, 방어, 보조 아이템들을 어떤 방법으로 분류하고 처리할지. 
	* 쿨타임 처리.

**Comment by LeeHyungJoo** [comment]
## 아이템 스킬 데이터 생각 및 설계

- 아이템 키에 종속되는가? 
사실 **아이템 레벨에 따라 다른 스킬을 쓰는 것을 궁극적으로 개발**했으면 좋겠음. 
즉, 스킬 테이블이 따로 있으면 좋을 것 같음. 
그 스킬 값에 따라 효과도 다르고, 걸리는 버프 및 효과들도 변경 가능하니. 

0. 사용 방식
a. ~선택 : 어떤 키를 누르면 주변 타겟들이 떠서 해당 타겟을 선택하여 스킬을 사용~
b. ~인디케이터 : 스킬 버튼을 누르고 조작하여 스킬 타겟을 잡고 사용. (논타겟도 가능)~
c. 자동으로 타겟을 잡아주거나, 스킬 사용시 현재 캐릭터 방향으로 시전을 한다

1. 스킬 시전 조건
a. ~사거리 : 0일 경우 사거리 판단 안 함 / 특정 값일 경우 2.a,b 에만 영향~
b. 쿨타임 : 0이면 오류. 

2. 타겟 선정 : 
표시 : 선정될 유닛들 아래에 원형 혹은 붉은 outline 세이더 사용

| 종류 | 범위 모양 | 범위 정보 1 | 범위 정보 2 | 이격 거리  |
|--------|--------|--------|--------|--------|
| 타겟 | -  | - | - |  - |  - |
| 위치 | 원형, 사각형 | r1 | r2 | - |
| 방향 | 원형, 사각형, 부채꼴 | r1 | r2 | 0이면 내 바로 앞에서 |
| 셀프 | 원형, 정사각형 | r1 | r2 | - |

?? 셀프 : 버프, 공전, 방어막 등등 

3. 캐스팅 시간
컬럼 값이 0이면 캐스팅 없고
특정 값일 경우,해당 값만큼 캐스팅.

4. 투사체 정보

+ 속도
   + 투사체 속도 0 일 경우 즉발 공격 
   + 특정 값일 경우, 해당 속도로 투사체 날라감. 

+ 충돌체 모양
   + 원형, 사각형

+ 충돌체 크기 정보
   + r1
   + r2

+ 유효 거리 
   + 0 이면 1.a 를 따름 
   + 특정 값일 경우, 해당 거리까지 날라간 후 소멸 

+ 만료 시간
   + 유효 거리를 우선적으로 판단.
   + 0이면 시간은 무제한
   + 특정 값일 경우, 일정 시간 뒤에 소멸 

5. 유도 여부 

+ 투사체의 속도가 0이 아닐 경우만 영향. 

6. 데미지 계산 여부 

+ FormulaType 시트 참조.

| 타입 | 유형 | 계산식 (다중 컬럼, 가중치만 입력|
|--------|--------|-------|
| 1 | Dealt | str 0.1 + def 0.2 ... |
| 2 | Heal | hp _0.2 ... |

**Comment by LeeHyungJoo** [comment]
## 아이템 성장 시스템 구현. 

0. 시트에 대한 질문
str,def,hp, cooltime 이 누적? 아니면 해당 값으로 대체 되는건가. 
쿨감은 퍼센트로 하는게 좋을 것 같음. 
그래야 후에 밸런스 잡기 편할 것 같음. 

1. 스택
경험치 양을 100~ 200 이런식 보다,
스택 단위로 하면 어떨까 생각해봤음. 
가시적으로, 아이템을 몇 개 먹어야, 레벨업이 되는지 쉽게 알수 있도록 
나중에 스택 하나를 뺏는다던지 ㅋㅋ 스킬 사용할때마다 효과 좋은 대신 스택 하나를 잃는다던지도 확장 개발 가능할 듯. 

[xxx][   ], [xx][xx][ ], [x][x][x][x][ ], [----MAX----]
1 -> 2 렙 2개 먹어야함. 
2 -> 3 렙 3개 먹어야함. 
3 -> 4 렙 5개 먹어야함. 등등

그런 후 MAX 일 경우 부턴 Gold 획득하는게 어떨까?
이건 서버에서 시스템 값 정해서 사용하겠음. 

2. 가이드
UpdateItemInfo 패킷으로 해당 값들을 관리. 


논의 후 일감으로 관리.

**Reply by KKangTaee** [reply]
골드 관련 시트도 개발예정

**Comment by LeeHyungJoo** [comment]
## 데이터에서 시간 표현
데이터의 경우 ms로 관리.

**Comment by LeeHyungJoo** [comment]
## 아이템 패킷 레이어 나누기.
```
  <struct name = "PItem">
    <int name ="itemKey"/>
  </struct>
```

```
	<struct name = "PInvenItem">
		<int name ="slot" />"
		<int name ="itemKey"/>
		<int name ="str"/>
		<int name ="def"/>
		<int name ="hp"/>
		<int name ="cooltime"/>
	</struct>
```

**Comment by LeeHyungJoo** [comment]
## 공격 가능 거리 들어왔을 경우 판단
```
        public static bool CanAttackDistance(in Entity self , in Entity target)
        {
            return self.currentPos.DistanceTo(target.currentPos) <= self.stat.AttackRange + target.stat.BodyRange;
        }
```
